### PR TITLE
Fix flaky test in RetryableS3OutputStreamTest

### PR DESCRIPTION
**Pull request by Akshat-Jain**
### Description

As part of https://github.com/apache/druid/pull/16481, we have started uploading the chunks in parallel. That means that it's not necessary for the `part that was uploaded at the last` to be less than or equal to the `chunkSize` (as the final part could've been uploaded earlier).

This made a test in RetryableS3OutputStreamTest flaky where we were asserting that the final part should be smaller than chunk size.

This PR fixes the test, and also adds another test where the file size is such that all chunk sizes would be of equal size.

<hr>

This PR has:

-  [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java [hunk 1]
```diff
@@ -171,6 +171,27 @@
...
+    final long fileSize = chunkSize * numChunks;
+    try (RetryableS3OutputStream out = new RetryableS3OutputStream(
+        config,
+        s3,
+        path,
+        s3UploadManager
+    )) {
+      for (int i = 0; i < fileSize; i++) {
+        out.write(i);
+      }
```
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java:187]: Made change.

File: extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java [hunk 2]
```diff
@@ -284,13 +304,15 @@ private void assertCompleted(long chunkSize, long expectedFileSize)
...
-        }
-      }
+      // Validate number of chunks uploaded.
+      final long numChunksOfChunkSize = expectedFileSize / chunkSize;
+      final int numChunksOfSmallerSize = expectedFileSize % chunkSize == 0 ? 0 : 1;
+      final long numOfExactChunks = partRequests.stream().filter(part -> part.getPartSize() == chunkSize).count();
+      final long numOfSmallerChunks = partRequests.stream().filter(part -> part.getPartSize() < chunkSize).count();
+
+      Assert.assertEquals(numChunksOfChunkSize, numOfExactChunks);
+      Assert.assertEquals(numChunksOfSmallerSize, numOfSmallerChunks);
```
Comment (kfaraz) [inline extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java:314]: Rephrase for readability:
```suggestion
      
      // Verify sizes of uploaded chunks
      int numSmallerChunks = 0;
      for (UploadPartRequest part : partRequests) {
          Assert.assertTrue(part.getPartSize() <= chunkSize);
          if (part.getPartSize() < chunkSize) {
              ++numSmallerChunks;
          }
      }
      Assert.assertTrue(numSmallerChunks <= 1)
```

File: extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java [hunk 3]
```diff
@@ -171,6 +171,26 @@ public void testWriteSmallBufferShouldSucceed() throws IOException
...
+  public void testWriteSmallBufferExactChunkSizeShouldSucceed() throws IOException
+  {
+    chunkSize = 128;
+    final int fileSize = 128 * 5;
+    try (RetryableS3OutputStream out = new RetryableS3OutputStream(
+        config,
+        s3,
+        path,
+        s3UploadManager
+    )) {
```
Comment (kfaraz) [inline extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java:184]: Nit: More readable like this:
```suggestion
    try (RetryableS3OutputStream out =
              new RetryableS3OutputStream(config, s3, path, s3UploadManager)) {
```

**Review by kfaraz — APPROVED** [review]

File: extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java (hunk 3 above)
Comment (Akshat-Jain) [reply extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java]: Done.

File: extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java (hunk 2 above)
Comment (Akshat-Jain) [reply extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetryableS3OutputStreamTest.java]: Done.

**Pull request merged**